### PR TITLE
Add support for VLAN QoS configuration

### PR DESCRIPTION
**Pull request by przemeklal**
* Add new `vlanQoS` field in the config spec.
* Call netlink library to set VLAN ID and VLAN QoS for a given VF interface.
* Reset VLAN ID and VLAN QoS configuration to default values on VF release.
* Update documentation.
* Update mocks.
* Use netlink library version that includes LinkSetVfVlanQos function.
* Update golang.org/x/sys to the version comptible with the updated netlink library version.

### CODE REVIEW

**Comment by przemeklal** [comment]
Hey guys, does the build fail because of the problem that the PR #93 solves? Interestingly it [passes](https://travis-ci.com/przemeklal/sriov-cni/builds/120191324) on my fork.

**Comment by zshi-redhat** [comment]
> Hey guys, does the build fail because of the problem that the PR #93 solves? Interestingly it [passes](https://travis-ci.com/przemeklal/sriov-cni/builds/120191324) on my fork.

I pulled the PR and passed the build/test locally, didn't get why it needs to update vendor dependency.
Above PR removes vendor dependency of build/lint/test rules in Makefile which may solve the build error.

**Comment by ahalim-intel** [comment]
Looks like anytime the glide files are updated the travis triggers `vendor`. I will merge #93. Maybe it's time to think about go module to mange dependencies.

**Comment by przemeklal** [comment]
Awesome, let me rebase on top of the lastest master and we'll see if that solves the problem.

File: pkg/sriov/sriov.go [hunk 1]
```diff
@@ -238,9 +244,14 @@ func (s *sriovManager) ApplyVFConfig(conf *sriovtypes.NetConf) error {
 	}
 
 	// 1. Set vlan
-	if conf.Vlan != 0 {
-		if err = s.nLink.LinkSetVfVlan(pfLink, conf.VFID, conf.Vlan); err != nil {
-			return fmt.Errorf("failed to set vf %d vlan: %v", conf.VFID, err)
+	if conf.Vlan != 0 || conf.VlanQoS != 0 {
```
Comment (ahalim-intel) [inline pkg/sriov/sriov.go]: The `VlanQoS` is depends on 'vlan` i.e. it's not possible to set vlan qos if vlan is not provided. I think you'll need to move the check for VlanQos inside  
```
if conf.Vlan != 0 {
    // 1. Set vlan
    // 2. Set vlan qos if it's non-zero
}
```
block.

Also, we need to check if vlanQos is given in the conf without vlan then plugin should return an error with appropriate message. You can do this check inside LoadConf. It's a good idea to document the usage in the configuration section too.

**Comment by zshi-redhat** [comment]
Should we reset `VlanQos` to `0` in `ResetVFConfig` just like `Vlan`? 
If `Vlan` is required to set `VlanQos`, can we remove `LinkSetVfVlan` method (`ResetVFConfig` is the only function using `LinkSetVfVlan`)?

**Comment by ahalim-intel** [comment]
> Should we reset `VlanQos` to `0` in `ResetVFConfig` just like `Vlan`?

Yes, we need to reset VlanQoS to `0` on Delete.

> If `Vlan` is required to set `VlanQos`, can we remove `LinkSetVfVlan` method (`ResetVFConfig` is the only function using `LinkSetVfVlan` after applying this PR)?

No, we still need this function to set the vlan. Not always we want to set vlan qos. Sometimes we my need to set only vlan and don't care about qos.

File: pkg/sriov/sriov.go (hunk 1 above)
Comment (przemeklal) [reply pkg/sriov/sriov.go]: Actually, that was my initial implementation, however later I found that `ip link` accepts non-zero vlan qos while vlan id is set to zero (e.g. `ip link set eth0 vf 0 vlan 0 qos 5`). Though on the i40e VFs the packets weren't tagged with the PCP information - I wasn't sure what to do and decided to keep things simpler.

I'll update this block to your proposed logic, thanks.

**Comment by przemeklal** [comment]
> Should we reset `VlanQos` to `0` in `ResetVFConfig` just like `Vlan`?

Ah, good catch, I updated the `resetVfVlan()` function to do that, however now I see that it's not even being called anywhere. Should I remove it then? Also, I'll update the `ResetVFConfig()` to reset both VLAN and QoS (however resetting VLAN ID to 0 seems to reset both already).

> If `Vlan` is required to set `VlanQos`, can we remove `LinkSetVfVlan` method (`ResetVFConfig` is the only function using `LinkSetVfVlan` after applying this PR)?

I can do that in this PR if there are no other calls to this function.

Thanks @zshi-redhat.

**Comment by zshi-redhat** [comment]
> > Should we reset `VlanQos` to `0` in `ResetVFConfig` just like `Vlan`?
> 
> Ah, good catch, I updated the `resetVfVlan()` function to do that, however now I see that it's not even being called anywhere. Should I remove it then? Also, I'll update the `ResetVFConfig()` to reset both VLAN and QoS (however resetting VLAN ID to 0 seems to reset both already).
> 

If `resetVfVlan()` is not used anywhere, I think we can remove it from the code.

**Comment by przemeklal** [comment]
> No, we still need this function to set the vlan. Not always we want to set vlan qos. Sometimes we my need to set only vlan and don't care about qos.

We actually may not need it. `LinkSetVfVlanQos` called with qos arg set to 0 acts exactly the same as `LinkSetVfVlan` and sends exactly the same netlink message to the kernel.

**Comment by ahalim-intel** [comment]
> Ah, good catch, I updated the `resetVfVlan()` function to do that, however now I see that it's not even being called anywhere. Should I remove it then? Also, I'll update the `ResetVFConfig()` to reset both VLAN and QoS (however resetting VLAN ID to 0 seems to reset both already).

You can remove the  `resetVfVlan()`, this is no longer in use.

> > If `Vlan` is required to set `VlanQos`, can we remove `LinkSetVfVlan` method (`ResetVFConfig` is the only function using `LinkSetVfVlan` after applying this PR)?
> 
> I can do that in this PR if there are no other calls to this function.

The reset should be done in  `ResetVFConfig()`.

**Comment by przemeklal** [comment]
I updated the logic - setting VLAN QoS now requires VLAN ID to be set as well. Reset moved to the appropriate function as discussed and the unused code removed.

I also updated the readme with information that vlanQoS requires vlan id or that the error is thrown if it's missing.

File: pkg/sriov/sriov.go [hunk 2]
```diff
@@ -272,6 +246,12 @@ func (s *sriovManager) ResetVFConfig(conf *sriovtypes.NetConf) error {
 
 	// Set vlan to 0
 	if conf.Vlan != 0 {
+		if conf.VlanQoS != 0 {
+			if err = s.nLink.LinkSetVfVlanQos(pfLink, conf.VFID, 0, 0); err != nil {
+				return fmt.Errorf("failed to set vf %d vlan: %v", conf.VFID, err)
+			}
+		}
+
 		if err = s.nLink.LinkSetVfVlan(pfLink, conf.VFID, 0); err != nil {
```
Comment (ahalim-intel) [inline pkg/sriov/sriov.go]: Should this be under an `else` block to avoid resetting vlan twice?

File: pkg/sriov/sriov.go [hunk 3]
```diff
@@ -244,14 +205,27 @@ func (s *sriovManager) ApplyVFConfig(conf *sriovtypes.NetConf) error {
...
+		return fmt.Errorf("non-zero vlan id must be configured to set vlan QoS")
+	}
 
-		if err = s.nLink.LinkSetVfVlanQos(pfLink, conf.VFID, conf.Vlan, conf.VlanQoS); err != nil {
-			return fmt.Errorf("failed to set vf %d vlan configuration: %v", conf.VFID, err)
+	if conf.Vlan != 0 {
+		// set vlan qos if present in the config
+		if conf.VlanQoS != 0 {
+			// validate that VLAN QoS is in the 0-7 range
+			if conf.VlanQoS < 0 || conf.VlanQoS > 7 {
```
Comment (ahalim-intel) [inline pkg/sriov/sriov.go]: I'd prefer to have this validation check in the config.LoadConf() and throw errors earlier than here. Once we get to  ApplyVFConfig() method we could safely assume that configs are valid and apply vlanqos if it's `Vlan` and `VlanQoS` both are present.
Just looking into  [ApplyVFConfig ](https://github.com/intel/sriov-cni/blob/01ab257c45960c6c6b1d4b9501bf7991fe16e9d7/pkg/config/config.go#L20) method and I see there's an opportunity to validate a vlan ids as well there if this is supplied.

File: pkg/sriov/sriov.go (hunk 2 above)
Comment (przemeklal) [reply pkg/sriov/sriov.go]: Yes, it should. I'll fix it.

File: pkg/sriov/sriov.go (hunk 3 above)
Comment (przemeklal) [reply pkg/sriov/sriov.go]: Good idea! I'll move it there.

Do you want me to include that VLAN ID range validation (1-4094) in this PR as well? Or leave it for now?

Comment (ahalim-intel) [reply pkg/sriov/sriov.go]: > Good idea! I'll move it there.
> 
> Do you want me to include that VLAN ID range validation (1-4094) in this PR as well? Or leave it for now?

That would be great. There are some other constraints on valid range though according to [cisco hw](https://www.orbit-computer-solutions.com/vlan-id-ranges-explained/) on a network. Please see [here]. But I don't think we need to go at that length right now. For now we can just validate against the above range.

Comment (przemeklal) [reply pkg/sriov/sriov.go]: Cisco VLAN constraints come from their proprietary ISL protocol. I'll go with the 0-4094 range then (1-4094 as valid range and 0 as default/disabled).

**Comment by przemeklal** [comment]
Rebased on top of the master branch after #88 merge. I also did a smoke test and it looks OK - configuration with VLAN ID, VLAN QoS and custom MAC address set at the same time worked fine.

**Comment by ahalim-intel** [comment]
@zshi-redhat Do you any further feedback on this PR?

**Comment by zshi-redhat** [comment]
> @zshi-redhat Do you any further feedback on this PR?

Just one minor question, when vlan equals to zero, can we still use vlanQos? from the code, we exclude this possibility, is it actually an option to use in reality?

**Comment by przemeklal** [comment]
Kernel accepts it, it shows up in the VF configuration, but 802.1Q header isn't inserted into packets and effectively packets are not tagged.

Please see this https://github.com/intel/sriov-cni/pull/94#discussion_r306722550.

**Comment by zshi-redhat** [comment]
> Kernel accepts it, it shows up in the VF configuration, but 802.1Q header isn't inserted into packets and effectively packets are not tagged.
> 
> Please see this [#94 (comment)](https://github.com/intel/sriov-cni/pull/94#discussion_r306722550).

Thanks for pointing to the comment, I should read that first.
Ok, it's configurable but not inserted in packet meaning other network devices cannot prioritize traffic based on the Qos settings.  
I'm fine with current code.

**Pull request merged**